### PR TITLE
Running crypten computation either with plans or jailing

### PR DESCRIPTION
**Pull request by youben11**
## Description
At the moment we have two ways of running crypten computations, either using plans or jails, instead of having two different branches that have similar functionalities, better to be able to use both in parallel.

This PR introduce a new message type CryptenInitJail to start a JailRunner on other syft workers, need to merge into syft-proto OpenMined/syft-proto#93. This way we can send messages to initiate either plans or jails.

## Type of change

Please mark options that are relevant.

- [ ] Added/Modified tutorials
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

* [x] I have added tests for my changes
* [x] I did follow the [contribution guidelines](https://github.com/OpenMined/PySyft/blob/master/CONTRIBUTING.md)
* [x] I have commented my code following [Google style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).

### CODE REVIEW

File: syft/messaging/message.py [hunk 1]
```diff
@@ -699,34 +699,99 @@ def contents(self):
...
+
+
+class CryptenInitJail(Message):
+    """Initialize a Crypten party using this message.
+
+    Crypten uses processes as parties, those processes need to be initialized with information
+    so they can communicate and exchange tensors and shares while doing computation. This message
+    allows the exchange of information such as the ip and port of the master party to connect to,
+    as well as the rank of the party to run and the number of parties involved. Compared to
+    CryptenInitPlan, this message also send two extra field, a JailRunner and a crypten model."""
```
Comment (LaRiffle) [inline syft/messaging/message.py:742]:
```suggestion
    CryptenInitPlan, this message also sends two extra fields, a JailRunner and a Crypten model."""
```

File: syft/messaging/message.py [hunk 2]
```diff
@@ -699,34 +699,99 @@ def contents(self):
...
+        """Return a human readable version of this message"""
+        return f"({type(self).__name__} {self.crypten_context})"
+
+    @property
+    def contents(self):
+        """Returns a tuple with the contents of the operation (backwards compatability)."""
+        return (self.crypten_context,)
+
+    @staticmethod
+    def simplify(worker: AbstractWorker, ptr: "CryptenInitJail") -> tuple:
```
Comment (LaRiffle) [inline syft/messaging/message.py:759]: Maybe rename ptr -> message ?

File: syft/messaging/message.py [hunk 3]
```diff
@@ -699,34 +699,99 @@ def contents(self):
...
+        Returns:
+            CryptenInitJail message.
+
+        Examples:
+            message = detail(sy.local_worker, msg_tuple)
+        """
+        msg_tuple = sy.serde.msgpack.serde._detail(worker, msg_tuple[0])
+        context = msg_tuple[0:4]
+        jail_runner = msg_tuple[4]
+        model = msg_tuple[5]
```
Comment (LaRiffle) [inline syft/messaging/message.py:796]:
```suggestion
        *context, jail_runner, model = msg_tuple
```

File: syft/workers/base.py [hunk 4]
```diff
@@ -457,6 +458,32 @@ def run_crypten_party(self, message: tuple):
...
+        ser_func = message.jail_runner
+        jail_runner = JailRunner.detail(ser_func)
+
+        rank = None
+        for r, worker_id in self.rank_to_worker_id.items():
+            if worker_id == self.id:
+                rank = r
+                break
+
+        assert rank != None
```
Comment (LaRiffle) [inline syft/workers/base.py:482]:
```suggestion
        assert rank is not None
```

File: syft/messaging/message.py [hunk 5]
```diff
@@ -699,34 +699,99 @@ def contents(self):
...
+    """Initialize a Crypten party using this message.
+
+    Crypten uses processes as parties, those processes need to be initialized with information
+    so they can communicate and exchange tensors and shares while doing computation. This message
+    allows the exchange of information such as the ip and port of the master party to connect to,
+    as well as the rank of the party to run and the number of parties involved. Compared to
+    CryptenInitPlan, this message also send two extra field, a JailRunner and a crypten model."""
+
+    def __init__(self, crypten_context, jail_runner, model=None):
+        self.crypten_context = crypten_context
```
Comment (LaRiffle) [inline syft/messaging/message.py:745]: I guess one example of Crypten context would be useful

File: syft/frameworks/torch/hook/hook.py [hunk 6]
```diff
@@ -141,7 +141,8 @@ def __init__(
         self._hook_native_tensor(torch.Tensor, TorchTensor)
 
         if dependency_check.crypten_available:
```
Comment (gmuraru) [inline syft/frameworks/torch/hook/hook.py:143]: Q: Should this be removed?

Comment (youben11) [reply syft/frameworks/torch/hook/hook.py]: Oh, I shouldn't have pushed that, I was only commenting this to not hook mpc tensor to get cat and stuff working, will just re-enable it for the moment

File: pip-dep/requirements.txt [hunk 7]
```diff
@@ -17,3 +17,4 @@ torch~=1.4.0
 tornado==4.5.3
 websocket_client~=0.57.0
 websockets~=8.1.0
+git+https://github.com/gmuraru/CrypTen.git@2aa0c2f14b472a8df6a85be1cfc4630479a10b2f#egg=crypten
```
Comment (youben11) [inline pip-dep/requirements.txt:20]:
```suggestion
git+https://github.com/facebookresearch/CrypTen.git#egg=crypten
```

File: syft/messaging/message.py [hunk 8]
```diff
@@ -699,34 +699,98 @@ def contents(self):
...
+            sy.serde.msgpack.serde._simplify(
+                worker, (*message.crypten_context, message.jail_runner, message.model)
+            ),
+        )
+
+    @staticmethod
+    def detail(worker: AbstractWorker, msg_tuple: tuple) -> "CryptenInitJail":
+        """
+        This function takes the simplified tuple version of this message and converts
+        it into an CryptenInitJail. The simplify() method runs the inverse of this method.
```
Comment (gmuraru) [inline syft/messaging/message.py:781]: TYPO: an -> a.

File: syft/workers/base.py [hunk 9]
```diff
@@ -453,11 +454,37 @@ def run_crypten_party(self, message: tuple):
                 rank = r
                 break
 
-        assert rank != None
+        assert rank is not None
```
Comment (gmuraru) [inline syft/workers/base.py:457]: Thank you for fixing this :D

**Pull request merged**